spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/annotation/SleuthAdvisorConfig.java

```diff
@@ -183,44 +181,141 @@ public Object invoke(MethodInvocation invocation) throws Throwable {
 		if (method == null) {
 			return invocation.proceed();
 		}
+
 		Method mostSpecificMethod = AopUtils
-				.getMostSpecificMethod(method, invocation.getThis().getClass());
+				.getMostSpecificMethod(invocation.getMethod(), invocation.getThis().getClass());
 		NewSpan newSpan = SleuthAnnotationUtils.findAnnotation(mostSpecificMethod, NewSpan.class);
 		ContinueSpan continueSpan = SleuthAnnotationUtils.findAnnotation(mostSpecificMethod, ContinueSpan.class);
 		if (newSpan == null && continueSpan == null) {
 			return invocation.proceed();
 		}
-		Span span = tracer().currentSpan();
-		if (newSpan != null || span == null) {
-			span = tracer().nextSpan().start();
+
+		Class<?> returnType = method.getReturnType();
+		if(isReactiveReturnType(returnType)){
+			return proceedUnderSpanReactively(invocation, newSpan, continueSpan);
+		} else {
+			return proceedUnderSpanSynchronously(invocation, newSpan, continueSpan);
+		}
+	}
+
+	private boolean isReactiveReturnType(Class<?> returnType) {
+		return Flux.class.equals(returnType) || Mono.class.equals(returnType);
+	}
+
+	private Object proceedUnderSpanSynchronously(
+			MethodInvocation invocation, NewSpan newSpan, ContinueSpan continueSpan) throws Throwable {
+		Span spanPrevious = tracer().currentSpan();
+		Span span;
+		if (newSpan != null || spanPrevious == null) {
+			span = tracer().nextSpan();
 			newSpanParser().parse(invocation, newSpan, span);
+			span = span.start();
+		} else {
+			span = spanPrevious;
 		}
 		String log = log(continueSpan);
 		boolean hasLog = StringUtils.hasText(log);
 		try (Tracer.SpanInScope ws = tracer().withSpanInScope(span)) {
```

Comment:
if spanPrevious is not null, this scoping is redundant

Reply:
fixed